Additional fields on vessel search

andres-arana opened the pull request:
Connects https://github.com/GlobalFishingWatch/GFW-Tasks/issues/887
Connects https://github.com/GlobalFishingWatch/GFW-Tasks/issues/993
Connects https://github.com/GlobalFishingWatch/GFW-Tasks/issues/1007
Connects https://github.com/GlobalFishingWatch/GFW-Tasks/issues/985
Connects https://github.com/GlobalFishingWatch/GFW-Tasks/issues/1050

The pipeline now has an additional step which connects to an elasticsearch server and pushes vessel information there. The query and schema for the import process are configurable via airflow configs, so this should be flexible enough to use on indo, ais and other data sources.

andres-arana committed:
Adds elasticsearch vessel info import process

andres-arana committed:
Renames elasticsearch field names for elasticsearch

andres-arana committed:
Adds changelog docs

andres-arana committed:
Fixes bad docs on how the elasticsearch auth setting is called

andres-arana committed:
Fixes broken dag flow

enriquetuya commented on assets/elasticsearch/vessels.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "mappings": {
+    "vessels": {
+      "properties": {
```
@andres-arana I think some fields are missing that should be present:
- SSVID
- Flag ISO-2

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,81 @@
+# Export records to json files
+################################################################################
+echo "Exporting records from $SOURCE"
+EXTRACT_PATH=$TEMP_PATH/bq/*.json
```
@andres-arana I think I preffer to have a `bq query` here instead of an extract, that will let us control a bit more what ends up being on the files. 

Also I think that it should be a pre-requisite that Elastcisearch receives the data on the needed format `YYYY-MM-DDT00:00:00Z`: https://github.com/GlobalFishingWatch/pipe-vessels/pull/1/files#diff-5973c79e28ebe22fefe322f6a76a9b04R18 and that should be achieved here in the `bq query`

enriquetuya commented on pipe_vessels/elasticsearch/importer.py:
```diff
@@ -0,0 +1,92 @@
+bq_datetime_regex = re.compile('^(\d\d\d\d-\d\d-\d\d) (\d\d:\d\d:\d\d) UTC$')
+
+
+def to_elasticsearch_datetime(value):
```
@andres-arana I think we need to make a pre-requisite that data comes in with the needed format for timestamps.

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,81 @@
+################################################################################
+echo "Loading data into Elastic Search"
+INDEX_SCHEMA=${ASSETS}/elasticsearch/vessels.json
+cat ${LOCAL_JSON_PATH}/*.json | python -m pipe_vessels.elasticsearch.importer $ELASTIC_SEARCH_SERVER_URL $ELASTIC_SEARCH_SERVER_AUTH $ELASTIC_SEARCH_INDEX_ALIAS $INDEX_SCHEMA
```
@andres-arana why not a `gsutil cat` instead of `cat`, that way you would avoid doing the download step?

enriquetuya replied:
@andres-arana I recall having some issues with logging when running python. I fixed this using `-u     : unbuffered binary stdout and stderr;` argument.

enriquetuya reviewed (changes requested):
@andres-arana check my comments. Let's discuss over slack if needed for the `bq query` one.

andres-arana committed:
Makes the vessel information query configurable

As requested on PR comments, and then on
https://github.com/GlobalFishingWatch/GFW-Tasks/issues/993, we need to
be able to configure different vessel information queries on each
pipeline run. This also requires the ElasticSearch index schema to be
confgiurable as well.

andres-arana committed:
Reimplements tracks and vessels import to cloudsql

andres-arana committed:
Changes the way the tracks are published

enriquetuya commented on airflow/post_install.sh:
```diff
@@ -14,7 +14,9 @@ python $AIRFLOW_HOME/utils/set_default_variables.py \
     bigquery_segment_vessel="segment_vessel" \
+    bigquery_segment_info="segment_info" \
     bigquery_tracks="tracks" \
+    bigquery_vessel_info="vessel_info" \
```
@andres-arana the last `\` on this line is not breaking anything?

andres-arana replied:
It doesn't break since it has an empty line after all of this. The idea here is that you can reorder these lines or add new lines without needing to change the last line, so that it doesn't show up in the diff. I've been doing this in all my pipelines so far.

enriquetuya commented on pipe_vessels/postgres/tracks.py:
```diff
@@ -1,16 +1,30 @@
+import json
+import csv
 
 def record_to_row(record):
```
Add some basic description of what this script does: `Loads the new line json of tracks and create csv so it can be imported by postgres.`

enriquetuya commented on scripts/aggregate_tracks:
```diff
@@ -3,24 +3,37 @@ source pipe-tools-utils
+################################################################################
 TABLE_DESC=(
   "* Pipeline: ${PIPELINE} ${PIPELINE_VERSION}"
   "* Source: ${MESSAGES_TABLE}"
```
Add that not only the meessages table is used but also: 
```
   -D segment_info=${SEGMENT_INFO_TABLE//:/.} \
   -D segment_vessel=${SEGMENT_VESSEL_TABLE//:/.} \
```

enriquetuya commented on scripts/aggregate_tracks:
```diff
@@ -29,43 +42,44 @@ TABLE_DESC=(
 echo "  Inserting new records for table ${TRACKS_TABLE}"
-
+INSERT_SQL=${ASSETS}/bigquery/aggregate-tracks.sql.j2
 jinja2 ${INSERT_SQL} \
```
Just a thought... I usually do this in a two step process:
```
SQL=jinja2 ${INSERT_SQL}
echo $SQL
echo $SQL | bq query
```
That way I have the query in the log for debugging purposes.

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,84 @@
+# Running bigquery query to generate vessel information
+################################################################################
+echo "Running vessel information query"
+EXTRACT_TABLE="${TEMP_EXTRACT_TABLE_PREFIX}_${CURRENT_TIMESTAMP}"
```
IMHO I would try to always perform queries in the same format in our bash so they are easier to read:
`echo $SQL | bq query`

enriquetuya replied:
I would change variable name to `TEMP_EXTRACT_TABLE`

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,84 @@
+echo "Running vessel information query"
+EXTRACT_TABLE="${TEMP_EXTRACT_TABLE_PREFIX}_${CURRENT_TIMESTAMP}"
+bq query \
+  --destination_table ${EXTRACT_TABLE} \
```
I would add `--headless` and `--allow_large_results` to `bq query`. This may work for indo, but may not scale for others.

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,84 @@
+  --destination_table ${EXTRACT_TABLE} \
+  --use_legacy_sql=false \
+  --replace \
+  "${SOURCE_QUERY}"
```
Add validation if the query succeeded or not
```
if [ "$?" -ne 0 ]; then
  echo ...
  exit 1
fi
```

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,84 @@
+## Export records to json files
+#################################################################################
+echo "Exporting records from $EXTRACT_TABLE"
+EXTRACT_PATH=$TEMP_PATH/bq/*.json
```
Change variable name to `TEMP_EXTRACT_PATH`

enriquetuya commented on scripts/aggregate_tracks:
```diff
@@ -29,43 +42,44 @@ TABLE_DESC=(
    -D dest=${TRACKS_TABLE//:/.} \
+   -D segment_info=${SEGMENT_INFO_TABLE//:/.} \
+   -D segment_vessel=${SEGMENT_VESSEL_TABLE//:/.} \
    | bq query --max_rows=0
```
I would add `--headless` and `--allow_large_results` to `bq query`. This may work for indo, but may not scale for others.

andres-arana replied:
I didn't know about the headless parameter The `allow_large_results` one is unnecessary if you are using standard SQL though:

`When specified, enables large destination table sizes for legacy SQL queries.`

enriquetuya commented on scripts/publish_vessel_info:
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
```
Add the `set -e`, we are validating if every step fails or not, just in case in the future we add some line and we forget to add the validation.

enriquetuya commented on assets/bigquery/aggregate-tracks.sql.j2:
```diff
@@ -1,112 +1,50 @@
-    ST_GEOGFROMTEXT(CONCAT( 'MULTIPOINT', " (", points_wkt, ')')) AS seg_geography
-  FROM
-    segments )
+    sv.vessel_id = '464b18bc4-404b-8298-3a5c-556f7d6c6453'
```
Debugging line, remove the `vessel_id` and the posterior `AND`

enriquetuya commented on assets/bigquery/aggregate-tracks.sql.j2:
```diff
@@ -1,112 +1,50 @@
 #standardSQL
```
@andres-arana this is okay for a first approach but we need to review this query. Especially talking about the `{{ source }}*`.  For indo only this will consume 8.5GB per day, for AIS, just for `201901` it is `160GB`. We need to split this query to add incremental data. 

```
#standardSQL
WITH
  ########################################################
  # Extract the messages we are going to process
  ########################################################
  messages AS (
  SELECT
    sv.seg_id,
    sv.vessel_id,
    m.timestamp,
    ST_GEOGPOINT(m.lon, m.lat) AS position,
    IFNULL(m.nnet_score, m.logistic_score) AS score,
    m.speed,
    m.course
  FROM
    `pipe_production_b.messages_scored_201901*` AS m
  INNER JOIN
    `pipe_production_b.segment_info` AS si
  USING
    (seg_id)
  INNER JOIN
    `pipe_production_b.segment_vessel` AS sv
  USING
    (seg_id)
  WHERE
    si.noise = false
    AND si.pos_count > 5
    AND m.lat IS NOT NULL
    AND m.lon IS NOT NULL
    AND sv.vessel_id_rank = 1)
SELECT
  seg_id,
  vessel_id,
  timestamp,
  position,
  score,
  speed,
  course
FROM
  messages
```

andres-arana replied:
Just for reference, this is in a separate ticket at https://github.com/GlobalFishingWatch/GFW-Tasks/issues/1050

andres-arana replied:
Actually, I ended up implementing this here as well, since we need this for the indonesia backfill.

enriquetuya commented on README.md:
```diff
@@ -30,16 +30,21 @@ The pipeline exposes the following standard settings:
-* `pipe_vessels.bigquery_tracks`: BigQuery table to write aggregated tracks into. Defaults to `tracks`
+* `pipe_vessels.source_table`: BigQuery table to read scored messages from, used as the source for track aggregation. Defaults to `messages_scored_`.
+* `pipe_vessels.bigquery_segment_vessel`: BigQuery table to read segment to vessel mappings. Defaults to `segment_vessel`.
+* `pipe_vessels.bigquery_tracks`: BigQuery table to write aggregated tracks into. Defaults to `tracks`.
```
@andres-arana mention here which is input and which is output table.

andres-arana replied:
@enriquetuya It's already saying that, "BigQuery table to **read** messages from", "BigQuery table to **write** aggregated tracks into, etc.

enriquetuya reviewed (changes requested):
@andres-arana please check my comments below.

andres-arana committed:
Implements changes due to PR review

andres-arana committed:
Makes the pipeline faster by importing only a day at a time

andres-arana committed:
Makes the whole process depend on past

andres-arana committed:
Removes yearly mode

andres-arana committed:
Index tuning

andres-arana committed:
It might be faster to just insert the daily records one by one with the indexes on than reindex on every copy
